pandas/tests/indexes/test_interval.py

```diff
@@ -326,170 +333,315 @@ def test_get_item(self):
+
+        index = IntervalIndex.from_tuples(
+            [(0, 2.5), (1, 3), (2, 4)], closed='right')
+
+        # single queries
+        queries = [Interval(1, 3, closed='right'),
+                   Interval(1, 3, closed='left'),
+                   Interval(1, 3, closed='both'),
+                   Interval(1, 3, closed='neither'),
+                   Interval(1, 4, closed='right'),
+                   Interval(0, 4, closed='right'),
+                   Interval(1, 2, closed='right')]
+        expected = [1, -1, -1, -1, -1, -1, -1]
+
+        for query, expected_result in zip(queries, expected):
+            result = index.get_indexer([query])
+            expect = np.array([expected_result], dtype='intp')
+            tm.assert_numpy_array_equal(result, expect)
+
+        # multiple queries
+        queries = [
+            [Interval(2, 4, closed='right'), Interval(1, 3, closed='right')],
+            [Interval(1, 3, closed='right'), Interval(0, 2, closed='right')],
+            [Interval(1, 3, closed='right'), Interval(1, 3, closed='left')],
+            index]
+        expected = [[2, 1], [1, -1], [1, -1], [0, 1, 2]]
+
+        for query, expected_result in zip(queries, expected):
+            result = index.get_indexer(query)
+            expect = np.array(expected_result, dtype='intp')
+            tm.assert_numpy_array_equal(result, expect)
+
+    @pytest.mark.xfail(reason="new indexing tests for issue 16316")
+    def get_indexer_for_ints_and_floats_updated_behavior(self):
+
+        index = IntervalIndex.from_tuples(
+            [(0, 1), (1, 2), (3, 4)], closed='right')
+
+        # single queries
+        queries = [-0.5, 0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5]
+        expected = [-1, -1, 0, 0, 1, 1, -1, -1, 2, 2, -1]
+
+        for query, expected_result in zip(queries, expected):
+            result = index.get_indexer([query])
+            expect = np.array([expected_result], dtype='intp')
+            tm.assert_numpy_array_equal(result, expect)
+
+        # multiple queries
+        queries = [[1, 2], [1, 2, 3], [1, 2, 3, 4], [1, 2, 3, 4, 2]]
+        expected = [[0, 1], [0, 1, -1], [0, 1, -1, 2], [0, 1, -1, 2, 1]]
+
+        for query, expected_result in zip(queries, expected):
+            result = index.get_indexer(query)
+            expect = np.array(expected_result, dtype='intp')
+            tm.assert_numpy_array_equal(result, expect)
+
+    @pytest.mark.xfail(reason="new indexing tests for issue 16316")
+    def get_indexer_nonunique_for_ints_and_floats_updated_behavior(self):
+
+        index = IntervalIndex.from_tuples(
+            [(0, 2.5), (1, 3), (2, 4)], closed='left')
+
+        # single queries
+        queries = [-0.5, 0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5]
+        expected = [[-1], [0], [0], [0, 1], [0, 1],
+                    [0, 1, 2], [1, 2], [2], [2], [-1], [-1]]
+
+        for query, expected_result in zip(queries, expected):
+            result = index.get_indexer_nonunique([query])
```

Comment:
Sure, please do open another issue to discuss. To be clear: I don't understand the exact motivation here, either, so we'll want to hear from Jeff on this one. One thing to consider is that unstructured lists of lists like `[[0, 1], [3, 4], [-1]]` can't be made used very efficiently, because they can't be represented by numpy arrays.